src/modsim2/similarity/metrics/pad.py

```diff
@@ -0,0 +1,322 @@
+        self.__train_data = None
+        self.__train_labels = None
+        self.__test_data = None
+        self.__test_labels = None
+        self.__embed_train_data = None
+        self.__embed_test_data = None
+        self.__models = []  # List of trained classifiers
+        self.__errors = []  # List of errors for classifiers on test data
+
+    def get_test_labels(self) -> np.ndarray:
+        """
+        This method is required for one of the tests - not used in the calculation
+        of the metric
+
+        Returns
+            self.__test_labels - the labels for the test datasest (a combination of
+            both A & B data)
+        """
+        return self.__test_labels
+
+    def __train_test_split(
+        self,
+        data_A: np.ndarray,
+        data_B: np.ndarray,
+    ):
+        """
+        This method splits the two datasets data_A and data_B into train and test
+        samples. To calculate the PAD, there should be the same number of samples
+        from A & B in the test dataset. However, there may not be the same number
+        of samples in A & B. The following lines of code determine the size of the
+        test dataset for both A & B - the minimum proportion size of the two datasets
+
+        Args:
+            data_A: all records in dataset A (excludes target value)
+            data_B: all records in dataset B (excludes target value)
+
+        Returns:
+            train_A: records in dataset A to be used for training
+            test_A: records in dataset A to be used for testing (final evaluation)
+            train_B: records in dataset B to be used for training
+            test_B: records in dataset B to be used for testing (final evaluation)
+        """
+        # Determine the number of samples of test data to be drawn from A&B
+        test_size_A = np.round(data_A.shape[0] * self.__test_proportion)
+        test_size_B = np.round(data_B.shape[0] * self.__test_proportion)
+        test_size = int(min(test_size_A, test_size_B))
+        # Split A&B into train and test
+        train_A, test_A = train_test_split(data_A, test_size=test_size, shuffle=True)
+        train_B, test_B = train_test_split(data_B, test_size=test_size, shuffle=True)
+
+        return train_A, test_A, train_B, test_B
+
+    @staticmethod
+    def __label_data(data: np.ndarray, label_value: int) -> np.ndarray:
+        """
+        Takes an array of data and an integer label value. Returns
+        an array of the label value with length matching the first
+        dimension of the array of data
+
+        Args:
+            data: array of data (any dimension)
+            label_value: integer value to set the label data
+
+        Returns:
+            labels: array of value label_value with length matching
+                    the number of data records
+        """
+        labels = np.ones(data.shape[0])
+        labels *= label_value
+        return labels
+
+    def __concat_data(
```

Comment:
more functional and imo clearer to have this return and the caller do the assignment to `self`

Reply:
i think applies elsewhere too. Happy to discuss whether worth changing as just stylistic

Reply:
I've got no strong views either way, have changed it as per your suggestion.